Network/Socket/Options.hsc

```diff
@@ -52,6 +52,12 @@ data SocketOption
     | UseLoopBack   -- ^ SO_USELOOPBACK
     | UserTimeout   -- ^ TCP_USER_TIMEOUT
     | IPv6Only      -- ^ IPV6_V6ONLY: don't use this on OpenBSD.
+    | RecvIPv4TTL      -- ^ Receiving IPv4 TTL.
+    | RecvIPv4TOS      -- ^ Receiving IPv4 TOS.
+    | RecvIPv4PktInfo  -- ^ Receiving IP_PKTINFO (struct in_pktinfo).
+    | RecvIPv6HopLimit -- ^ Receiving IPv6 hop limit.
+    | RecvIPv6TClass   -- ^ Receiving IPv6 traffic class.
+    | RecvIPv6PktInfo  -- ^ Receiving IPV6_PKTINFO (struct in6_pktinfo).
     | CustomSockOpt (CInt, CInt)
```

Comment:
Though it would be an API-breaking change, also the socket options SHOULD NOT be (have been) a sum type. This is an open type, with platform-specific code-points that change over time, users should be able to set socket options not known to the library.

Reply:
We can use `CustomSockOpt` for socket options not known to the library.

Reply:
At this point perhaps, if you want maintain legacy compatibility, though the interface is also too limited in the types of data that one can provide with a socket option. It needs to support not options that pass an integral type by value, but also options that pass other storable types by reference.

Still pattern synonyms are a much better fit for supporting an integral type partially covered by a variable set of pre-defined values.

Reply:
We should implement a new API, say, `setSockOpt` for the generic purpose while keep `setSocketOption` for backward compatibility.

Reply:
Yes, perhaps that's the right approach.